include/seqan3/alignment/scoring/hamming_scoring_scheme.hpp

```diff
@@ -0,0 +1,92 @@
+// SPDX-FileCopyrightText: 2006-2024 Knut Reinert & Freie Universität Berlin
+// SPDX-FileCopyrightText: 2016-2024 Knut Reinert & MPI für molekulare Genetik
+// SPDX-License-Identifier: BSD-3-Clause
+
+/*!\file
+ * \brief Provides seqan3::hamming_scoring_scheme.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <concepts>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3
+{
+
+/*!\brief A scoring scheme that assigns a score of 0 to matching letters and -1 to mismatching letters.
+ * \ingroup alignment_scoring
+ *
+ * \details
+ *
+ * This stateless scoring scheme is equivalent to the Hamming distance and assigns a score of 0 to matching letters and -1 to
+ * mismatching letters.
+ * This scheme is independent of the alphabet type and can be used whenever the two compared alphabets model the
+ * std::equality_comparable_with concept.
+ * Use this scheme if you want to use the use the more efficient bitparallel alignment algorithm often used in the
+ * context of computing the edit distance.
+ * See the documentation for \ref seqan3::align_cfg::edit_scheme for more details.
+ */
+class hamming_scoring_scheme
+{
+public:
+    //!\privatesection
+    //!\brief The alphabet type of the scoring scheme. This type is only used for the alignment configuration machinery.
+    using alphabet_type = char;
+    //!\publicsection
+
+    //!\brief The underlying score type.
+    using score_type = int32_t;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    hamming_scoring_scheme() noexcept = default; //!< Defaulted.
+    //!\}
+
+    /*!\name Accessors
+     * \{
+     */
+
+    /*!\brief Returns the score of two letters.
+     * \tparam alph1_t The type of the first letter.
+     * \tparam alph2_t The type of the second letter.
+     * \param[in] alph1 The first letter to compare.
+     * \param[in] alph2 The second letter to compare.
+     *
+     * \details
+     *
+     * This function returns 0 if the two letters are equal and -1 otherwise. Note both the alphabet types of the
+     * compared letters must model the std::equality_comparable_with concept, otherwise no overload of this function
+     * is available.
+     *
+     * \returns The score of the two letters. 0 if the letters are equal, -1 otherwise.
+     */
+    template <typename alph1_t, typename alph2_t>
+        requires std::equality_comparable_with<std::remove_reference_t<alph1_t>, std::remove_reference_t<alph2_t>>
```

Comment:
One could potentially call the interface with explicit template arguments so that the types could be reference types. But I also took a look into the concept definition and there the equality check is always performed on const lvalue-references no matter what type one would put in. So I guess less boilerplate is better for reading and removing the reference types are the way to go.